prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt

```diff
@@ -51,4 +58,108 @@ object FHIRBundleHelpers {
+        return properties.filter { it.hasValues() }.flatMap { it.values }
+            .filterIsInstance<Reference>().map { it.reference }
+    }
+
+    /**
+     * Gets all child properties for a resource [property] recursively
+     *
+     * @return a flatmap stream of all child properties on a [property]
+     */
+    private fun getChildProperties(property: Property): Stream<Property> {
+        return Stream.concat(
+            Stream.of(property),
+            property.values.flatMap { it.children() }.stream().flatMap { getChildProperties(it) }
+        )
+    }
+
+    /**
+     * Gets all diagnostic report that have no observations from a [bundle]
+     *
+     * @return a list of [Base] diagnostic reports that have no observations
+     */
+    fun getDiagnosticReportNoObservations(bundle: Bundle): List<Base> {
+        return FhirPathUtils.evaluate(
+            null,
+            bundle,
+            bundle,
+            "Bundle.entry.resource.ofType(DiagnosticReport).where(result.empty())"
+        )
+    }
+
+    /**
+     * Deletes a [resource] from a bundle, removes all references to the [resource] and any orphaned children.
+     * If the [resource] is a [DiagnosticReport] and it has no observations. The [DiagnosticReport] and all
+     * its references will be removed
+     *
+     */
+    fun Bundle.deleteResource(resource: Base) {
+
+        if (this.entry.find { it.fullUrl == resource.idBase } == null) {
+            throw IllegalStateException("Cannot delete resource. FHIR bundle does not contain this resource")
+        }
+
+        // Get the resource children references
+        val resourceChildren = resource.getResourceReferences()
+        // get all resources except the resource being removed
+        val allResources = this.entry.filter { it.resource.id != resource.idBase }
+        // get all references for every resource
+        val allReferences = allResources.flatMap { it.getResourceReferences() }
+
+        // remove orphaned children
+        resourceChildren.forEach { child ->
+            if (!allReferences.contains(child)) {
+                this.entry.removeIf { it.resource.id == child }
+            }
+        }
+
+        // Go through every resource and check if the resource has a reference to the resource being deleted
+        // if there is remove the reference
+        this.entry.forEach { res ->
+            res.resource.getResourceProperties().forEach { property ->
+                property.values.forEach {
+                    if (it is Reference && it.reference == resource.idBase) {
+                        it.reference = null
+                        it.resource = null
+                    }
+                }
+            }
+        }
+
+        // if resource being removed is an observation
+        // check if there are any diagnostic reports with no observations and remove them
+        if (resource is Observation) {
```

Comment:
Yeah, I don't like this either. I tried looking for a way to make it generic, but couldn't find it. If I make an extension function on Observation I still need to check if resource is Observation, otherwise it's just going to call the Base function. I guess it would work if we knew the type of the resource before calling deleteResource, but we don't. We can get on a call if I'm missing something.

Reply:
Hmm I'll try to take a look today and see if I can think of something.

Reply:
Can't really think of anything else right now, so I guess this is fine. If you wanted to, you could move this logic to a separate method at least and call it something like `deleteChildlessParentResource` or something and have a `when` clause for different types.

Reply:
Created a new function to separate the logic and added unit tests for it.